### PR TITLE
Nspec update

### PR DESCRIPTION
**Pull request by noodlemoodle**
- update nspec to use the authorizableauthorizer so that read is only allowed for unauthenticated nspecs and nspecs for which the user has permission "service.{id}" to

- reload the client on fetch subject success

nano https://github.com/nanoPayinc/NANOPAY/pull/14502

### CODE REVIEW

File: src/foam/nanos/boot/NSpec.js [hunk 1]
```diff
@@ -282,6 +283,78 @@ foam.CLASS({
...
+      javaThrows: ['AuthorizationException'],
+      javaCode: `
+        if ( ! getAuthenticate() ) return;
+
+        String permission = "service." + getId();
+        AuthService auth = (AuthService) x.get("auth");
+  
+        if ( ! auth.check(x, permission) ) {
+          ((foam.nanos.logger.Logger) x.get("logger")).debug("AuthorizableAuthorizer", "Permission denied.", permission);
+          throw new AuthorizationException("Permission denied: Cannot reda this NSpec.");
```
Comment (jlhughes) [inline src/foam/nanos/boot/NSpec.js]:
```suggestion
          throw new AuthorizationException("Permission denied: Cannot read this NSpec.");
```

File: src/foam/nanos/auth/CachedAuthServiceProxy.js [hunk 2]
```diff
@@ -32,7 +32,7 @@ foam.CLASS({
       if ( this.loginSuccess$ ) {
         this.onDetach(this.loginSuccess$.sub(this.resetCache));
       }
-      this.onDetach(this.userCapabilityJunctionDAO.on.sub(
+      this.onDetach(this.loginSuccess && this.userCapabilityJunctionDAO.on.sub(
```
Comment (kgrgreer) [inline src/foam/nanos/auth/CachedAuthServiceProxy.js]: Wouldn't it be better if this was:
this.loginSuccess && this.onDetach(...);

File: src/foam/nanos/boot/Boot.java [hunk 3]
```diff
@@ -156,7 +156,7 @@ public Object create(X x) {
     ((ProxyDAO) root_.get("nSpecDAO")).setDelegate(
       new foam.nanos.auth.AuthorizationDAO.Builder(getX())
         .setDelegate(serviceDAO_)
-        .setAuthorizer(new foam.nanos.auth.GlobalReadAuthorizer("service"))
+        .setAuthorizer(new foam.nanos.auth.AuthorizableAuthorizer("service"))
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java:159]: Couldn't we just use a StandardAuthorizer, and then avoid making the NSpec Authorizable? Is there any custom logic that is required?

Comment (noodlemoodle) [reply src/foam/nanos/boot/Boot.java:159]: the read uses custom logic which checks authenticate and a different permission from the standardauthorizer but the other methods do not need custom logic. ~created a pr to address https://github.com/kgrgreer/foam3/pull/1598~

**Pull request merged**